### ISSUE TITLE
[HOLD][$1000] Web- Preferences - The recent mode description is truncated in Spanish preference 

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#17509]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to Web chrome
2. Change language to Spanish
3. Go to preferences and mode
4. See that the description is truncated at the end

## Expected Result:
The recent description should not be truncated at the end and should be shown in two lines

## Actual Result:
The recent description is truncated at the end

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.0.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/232520168-0e30024c-0b61-4151-8ecf-f22f0fe1ccfc.mp4

![image (33)](https://user-images.githubusercontent.com/93399543/232520174-e5011029-4526-4e95-ad80-0cea034a39b5.png)

**Expensify/Expensify Issue URL:**

**Issue reported by:** @avi-shek-jha

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681621048024679

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01decf2433726a790f</li>
        <li>Upwork Job ID: 1648665877492830208</li>
        <li>Last Price Increase: 2023-05-03</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @flaviadefaria (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01decf2433726a790f

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @s77rt (`External`)

**Comment by dukenv0307** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The recent mode description is truncated in Spanish preference
### What is the root cause of that problem?
Option of `priorityModes` have property `alternateTextMaxLines` is undefined.
https://github.com/Expensify/App/blob/4019be1c2d41bb27bab90544a84dff7c502f1d3c/src/pages/settings/Preferences/PriorityModePage.js#L41
Then, style `white-space: "pre"` is applied. So that when text description is too long it will be truncated.
https://github.com/Expensify/App/blob/4019be1c2d41bb27bab90544a84dff7c502f1d3c/src/components/OptionRow.js#L128
### What changes do you think we should make in order to solve the problem?
Change `alternateTextMaxLines` to `2`
https://github.com/Expensify/App/blob/4019be1c2d41bb27bab90544a84dff7c502f1d3c/src/pages/settings/Preferences/PriorityModePage.js#L41
### What alternative solutions did you explore? (Optional)
**NA**
### Result 

[Screencast from 18-04-2023 16:53:06.webm](https://user-images.githubusercontent.com/129500732/232741652-7baffac1-bce8-4ff8-b3bc-f5a1fcfffa22.webm)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @marcochavezf (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by bernhardoj** [comment]
As @allroundexperts mentioned and this comment https://github.com/Expensify/App/pull/15130#issuecomment-1498022925, `numberOfLines` 0 with `lineHeight` style will throw an error on Android, but this only happens on our fork. It could be means that something missed when we are upgrading the react native. I think we should investigate that.

**Comment by s77rt** [comment]
@dukenv0307 Thanks for the proposal. Your RCA seems about right. But I think the fix can be improved. We want `alternateTextMaxLines` to take as many lines as needed and not just limited to 2.

**Comment by s77rt** [comment]
@bernhardoj Right, this is a regression from https://github.com/Expensify/App/pull/15130 (https://github.com/Expensify/App/commit/bdf28204faa48f783ef14b599cc29af69fc9c898). We should have kept `alternateTextMaxLines` set to 0 not to restrict the number of lines. Setting `alternateTextMaxLines` to `undefined` is basically the same as setting it to 1 since we use `lodashGet` [here](https://github.com/Expensify/App/blob/5c7b7c46501c80e6d99bd2119bb1e01a17378fc8/src/components/OptionRow.js#L248) where we fallback to 1.

cc @priyeshshah11 @aimane-chnaif @roryabraham

**Comment by s77rt** [comment]
@allroundexperts Oh so sorry, I didn't see the proposal it was before `Help Wanted` label. I used to only review proposals below that :sweat_smile: Checking now...

**Comment by s77rt** [comment]
@allroundexperts Thanks for the proposal. I agree with you on the RCA. But I don't think setting `alternateTextMaxLines` to 2 will crash the App, will it? If not then that would be our backup plan.

We can't really keep running from fixing the RN bug especially since it only happens on our fork (based on @bernhardoj comment). What I'm looking for here is to be able to set `alternateTextMaxLines` back to `0` and fix the Android crash.

**Comment by dukenv0307** [comment]
@s77rt I think if we want to improve `alternateTextMaxLines` we need to set value to it in `OptionRow` depend on `this.props.option.alternateText` and with of wrap content

**Comment by s77rt** [comment]
@dukenv0307 We want to set it to 0 and fix the Android crash.

**Comment by s77rt** [comment]
@allroundexperts I think the bug only exists on our fork and not on RN. I'm not sure if this was ever a bug upstream that was fixed or it's just a bug that was introduced in our fork.

**Comment by dukenv0307** [comment]
@s77rt I tried to test this bug in main branch of my fork after sync fork. I used correctly version RN now, it will crash in Android when set `alternateTextMaxLines` to 0, but it's not happen when I reverted version RN to `v0.71.2-alpha.0`

**Comment by s77rt** [comment]
@dukenv0307 Thanks for the confirmation. Do you have any suggestion on how we can fix the Android crash?

**Comment by dukenv0307** [comment]
I think the root problem of the Android crash is this [PR](https://github.com/Expensify/react-native/pull/37/files) when apply numberOfLines and maximumNumberOfLines props on iOS and Android

**Comment by bernhardoj** [comment]
Sharing my finding.

Yes, the error comes from this PR https://github.com/Expensify/react-native/pull/37, specifically this changes:
https://github.com/Expensify/react-native/pull/37/files#diff-3067d6082d9bb646602ef7588f03fc4cc294a65341042f83a4a54afcd4d01dfd

where the NUMBER_OF_LINES is changed to -1.

The crash comes from this code
<img width="376" alt="image" src="https://user-images.githubusercontent.com/50919443/233640865-2dc9cafb-853a-4e61-8a1a-851acbe65a1e.png">
TOP value is 1
<img width="322" alt="image" src="https://user-images.githubusercontent.com/50919443/233640971-b0dabcf1-8b0d-4e44-a14e-67eee5bc0c2e.png">
and mColumns value is COLUMNS_NORMAL (5)
<img width="369" alt="image" src="https://user-images.githubusercontent.com/50919443/233641217-36b44fd4-c9b0-4dcb-b79c-89542736b368.png">

which makes the number
5 * -1 + 1 = -4
the same number we got from the crash

and looks like the call comes from this code
<img width="638" alt="image" src="https://user-images.githubusercontent.com/50919443/233641951-f9108a79-c025-4b82-8c25-6fff3eb4dde7.png">

if numberOfLines is 0, we take the default value which is -1. I tried to change it and other occurrences to ViewDefaults.MAX_NUMBER_OF_LINES. It doesn't crash, but the numberOfLines doesn't work anymore 🙃.

Maybe the PR author @Szymon20000 can help us.

**Comment by Szymon20000** [comment]
Sure, I forgot to update few places in the Text component and somehow In RNtester I wasn’t able to spot a problem. Will do my best to fix it quickly.

**Comment by Szymon20000** [comment]
Hey!
I think this one should fix it. https://github.com/facebook/react-native/pull/35703/commits/b58d91490f0b23798e38cd29c807d04c6a7e6fc9

**Comment by s77rt** [comment]
@Szymon20000 Thank you. Can you please post a proposal following the [proposal template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md)?

**Comment by flaviadefaria** [comment]
Waiting on a final proposal to move this forward.

**Comment by flaviadefaria** [comment]
Friendly bump @Szymon20000!

**Comment by Szymon20000** [comment]
To be honest not sure if my proposal is needed here. 
Someone pointed that the issue is causes by my recent pr (https://github.com/Expensify/react-native/pull/37.
I update my upstream pr in react-native repository and the fix will be available from the next expensive RN fork release.

**Comment by s77rt** [comment]
@Szymon20000 I think the proposal is still needed for us to get an idea on what went wrong and how we are fixing the bug. I understand that you made the first changes in the first PR and fixed that recently but still we can't merge blindly. I would still have to review the PR unless it was already reviewed and merged upstream. Is it the case here?

**Comment by flaviadefaria** [comment]
@Szymon20000 would it be too much trouble to get the proposal in? We'd really love to keep making progress here.

**Comment by Szymon20000** [comment]
Sorry @flaviadefaria @s77rt  I was super busy last few days. 
Let me write it now if that's helpful

**Comment by Szymon20000** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
App is crashing or not showing views within nested text components
### What is the root cause of that problem?
I created a pr that was supposed fix numberOfLines prop on android and add support for iOS  in TextInput component. However, it turned out that Text component also supports the prop but in an inconsistent way. In the Text component numberOfLines={3} would actually mean that we want to truncate the text if it has more than 3 lines. However, if the text has 2 lines we don't increate the height to 3. So it behaved as maximumNumberOfLines. I made it consistent by renaming it on the native side as maximumNumberOfLines as some parts of the implementation is shared between TextInput and Text components. I also changed the default value for numberOfLines props as it was no longer in used in the Text component. However it turned out that even though it was no longer used in Text shadow node and textLayoutManager it was still used in some parts of the ReactTextComponent.java as a result it wasn't working properly.
### What changes do you think we should make in order to solve the problem?
Just make Text Input component to use previous default value which is now set in ViewDefaults.MAXIMUM_NUMBER_OF_LINES
here is the full pr: https://github.com/facebook/react-native/pull/35703/commits/b58d91490f0b23798e38cd29c807d04c6a7e6fc9

AS we recently finally got a review from a RN contributor I think we will publish our fork without my changes and once the pr is merged in upstream repo we will publish another version with accepted solution

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by s77rt** [comment]
@Szymon20000 Thanks for the proposal. Can you please explain why the change of `numberOfLines` prop on Text / TextInput was necessary. Wasn't the main goal of the PR the addition of `maxNumberOfLines` prop for both Android and iOS? I understand that would keep `numberOfLines` essentially the same as `maxNumberOfLines` on Text but that is the current behaviour, maybe this should have been fixed by a follow up PR instead.

Checking the linked PR https://github.com/facebook/react-native/pull/36394 mapping `numberOfLines` to `maximumNumberOfLines` does not feel right. Those are different props that do different things, why we are mapping one to another?

___

@roryabraham Do you think we should revert https://github.com/Expensify/react-native/pull/37? That PR should not have been included in the latest version since it's incomplete, there is a complementary PR upstream https://github.com/facebook/react-native/pull/36394 but it's not ready yet.

We are also getting some reviews on the initial upstream PR here https://github.com/facebook/react-native/pull/35703.

**Comment by Szymon20000** [comment]
@s77rt It does feel right.
The goals was actually to implement numberOfLines for textInput but I also added maxNumberOfLines as android and Ios work asynchronously in comparison to web. Please read the rn docs to check that numberOfLines for TextInput and Text behaved differently. Because both components on the native side in new arch shares the same code for computing the layout I had to make it consistent. NumberOfLines on the Text component should only set a maximum Limit of lines while in Text Input ot should set the exact height in lines. Please check docs for TextInput for android. It also uses setLines on the native side that should do exactly that.

**Comment by s77rt** [comment]
@Szymon20000 I understood that. What I'm trying to say is that I don't see why we are mapping props.

## Current (RN):
### Text:
- numberOfLines: :x: This is actually maxNumberOfLines
- maxNumberOfLines:  :x: Not implemented

### TextInput
- numberOfLines: :heavy_check_mark: 
- maxNumberOfLines:  :x: Not implemented

## Your PR:
### Text:
- numberOfLines: :x: This is still maxNumberOfLines and internally you are basically mapping this value to maxNumberOfLines and making numberOfLines a dead code. However this was actually used somewhere else where we did not map it correctly thus causing the bug.
- maxNumberOfLines: :heavy_check_mark:

### TextInput
- numberOfLines: :heavy_check_mark: 
- maxNumberOfLines: :heavy_check_mark:


## Suggestion:
### Text:
- numberOfLines: :grey_question: I still don't understand why you had to modify this, isn't the new behaviour exactly the same as it was before? This is lacking on iOS I understand that, but why making change to Android if the final results are the same?
- maxNumberOfLines: :heavy_check_mark:

### TextInput
- numberOfLines: :heavy_check_mark: 
- maxNumberOfLines: :heavy_check_mark:

**Comment by Szymon20000** [comment]
Hey @s77rt !
Thanks for spending your time on it!
So the reason for renaming it is that numberOfLines and maximumNumberOfLines has different names in Fabric but the same native C++ component is reused in TextComponent and TextInputComponent. To avoid runtime checks in Js code We decided to go this way. Let me send you few links. 
https://github.com/facebook/react-native/blob/714b502b0c7a5f897432dbad388c02d3b75b4689/packages/react-native/ReactCommon/react/renderer/components/text/ParagraphProps.cpp#L77 <- before my changes
https://github.com/facebook/react-native/blob/714b502b0c7a5f897432dbad388c02d3b75b4689/packages/react-native/ReactCommon/react/renderer/components/textinput/androidtextinput/react/renderer/components/androidtextinput/AndroidTextInputProps.cpp#L55 <- before my changes. 
Both Text and TextComponent uses the same class for layouting on Fabric and the same structure for sending text attributes from C++ to Java. in order to avoid coping all of that I had to rename it. 
You may ask why it worked before on android. Why we can't do what was there before. So in Fabric they changed behaviour of numberOfLines in textInput to the same as in Text which doesn't make sense as on React-native for web we set fixed number of lines in TextArea and want to have consistent behaviour across platforms.

So we actually have 3 goals here: 
- make fabric and paper consistent
- make Android, iOS, consistent
- As Text and TextComponent reuses a lot of native classes (Make them consistent too (use maxNumberOfLines in Text component) as that's what it's actually doing )

**Comment by s77rt** [comment]
@Szymon20000 Thanks a lot for the follow up and quick responses!
@roryabraham Would you please hold-assign @Szymon20000

___

@Szymon20000 Any ETA on the linked complementary PR? Also I see some requested changes on the initial PR, would you please resolve those as well

**Comment by s77rt** [comment]
@roryabraham ^

**Comment by flaviadefaria** [comment]
@roryabraham @marcochavezf what are your thoughts on this one?

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by marcochavezf** [comment]
Thanks @aimane-chnaif and @s77rt for pointing out the related PRs. Let's put this issue on hold in favor of https://github.com/facebook/react-native/pull/35703 and https://github.com/facebook/react-native/pull/36394. After those PRs are merged, we can re-evaluate the situation to determine if the problem still persists or has been resolved.

**Comment by bernhardoj** [comment]
If we are not going with the patch here https://github.com/Expensify/App/issues/17509#issuecomment-1525753477, I would say we revert the PR https://github.com/Expensify/react-native/pull/37 to allow `numberOfLines` 0 and solve these issues https://github.com/Expensify/App/issues/17441 https://github.com/Expensify/App/issues/17368 https://github.com/Expensify/App/issues/17433 which I think is really bad if we keep those issues unsolved for an indefinite amount of time.

**Comment by mwmsh** [comment]
> Thanks @aimane-chnaif and @s77rt for pointing out the related PRs. Let's put this issue on hold in favor of [facebook/react-native#35703](https://github.com/facebook/react-native/pull/35703) and [facebook/react-native#36394](https://github.com/facebook/react-native/pull/36394). After those PRs are merged, we can re-evaluate the situation to determine if the problem still persists or has been resolved.

How does this affect #17441? Has anyone verified?

**Comment by s77rt** [comment]
Not overdue.

@marcochavezf Shouldn't this be on hold already :sweat_smile:? Or are we planning to revert https://github.com/Expensify/react-native/pull/37?

**Comment by marcochavezf** [comment]
Oops I forgot to apply the hold label here

**Comment by s77rt** [comment]
Still on hold

**Comment by flaviadefaria** [comment]
I'll switch this to weekly while it's on hold. If anyone removes it from hold, please also switch the priority to daily.

**Comment by s77rt** [comment]
Same ^

**Comment by marcochavezf** [comment]
> On hold

**Comment by marcochavezf** [comment]
On hold

**Comment by marcochavezf** [comment]
Still on hold

**Comment by marcochavezf** [comment]
Still on hold

**Comment by marcochavezf** [comment]
Still on hold

**Comment by s77rt** [comment]
@marcochavezf @flaviadefaria This seems fixed already. I think we can close this? or pay the bug reporter and close

**Comment by marcochavezf** [comment]
Do you know by chance if this issue was reported in another GH issue?

**Comment by s77rt** [comment]
Can we close this?

**Comment by avi-shek-jha** [comment]
Will I get paid for this?

**Comment by avi-shek-jha** [comment]
> This is fixed in #18507

@flaviadefaria I see that the other issue was created after my issue. So I guess, I should be the initial reporter for this issue. Thanks.
